docker-compose.yml

```diff
@@ -0,0 +1,12 @@
+version: '3'
+services:
+  frontend:
+    build: ./frontend
+    volumes:
+      - ./frontend:/usr/local/src
+      - node_modules:/usr/local/src/node_modules
+    working_dir: /usr/local/src
+    entrypoint: yarn
+    command: start
+volumes:
+  node_modules:
```

Comment:
The named module here helps prevent `node_modules` from getting overwritten by the host OS when we mount `./frontend:/usr/local/src`.

Reply:
For my own knowledge! Could you say more about why `node_modules` would be overwritten, otherwise? Or perhaps point me to a useful resource?

Reply:
Sure! The gist of it is that `node_modules` is a directory that gets created automatically by `npm isntall`/`yarn install` to hold dependencies (kind of like `site-packages` in Python) and it needs to live in the application directory in order to be importable into the application code. There are two ways to manage it in a container:

1. Run `npm install` or `yarn install` in the Dockerfile as a build step, meaning that `node_modules` is pre-buillt in the application container image and doesn't exist on the host OS
2. Run `docker-compose run frontend install` after the container image has already been built, meaning that `node_modules` will get installed on the host OS and then passed into the container when the application directory is mounted as a volume

In my experience 1) is preferable to 2), since NPM packages [are sometimes different for different operating systems](https://stackoverflow.com/a/35238945), and the Docker host OS is often different from the guest OS. So if you install packages for the host OS and then pass them to the container, sometimes you wind up with the wrong package for the container OS.

But 1) comes with the snag that when you mount the application directory from the host OS to the container (which you need to do in order for your local changes to be visible in the container) it will overwrite that directory in the container. Since 1) installs the `node_modules` in the container and not on the host OS, mounting the directory from the host OS to the container will overwrite the dependencies.

The named volume gets around this by creating a "third" location on disk where the volume will be stored. This means we can safely mount `frontend` to `/usr/local/src/` in the container without worrying about `usr/local/src/node_modules` getting overwritten, because `/usr/local/src/node_modules` is actually linked to the third location where the named volume is stored.

For more on the different types of volumes, see: https://success.docker.com/article/different-types-of-volumes Unfortunately I haven't found good docs on this particular solution to the problem, but it's what I've cobbled together after asking a few friends that use Docker and NPM frequently!

Reply:
Wow, thanks! 

In sum...the Dockerfile builds `node_modules` into the container. But (oh no!) we still have to mount the application directory to the container (to makes changes present in the container itself): this essentially undoes the work of `RUN yarn install`. So, we sneakily store the `node_modules` elsewhere – a named volume (`node_modules`) linked to `/usr/local/src/node_modules`. Very clever.